### ISSUE TITLE
output looks the same

### ISSUE THREAD

**Issue by BustedSec** [anhlqn/elastic#1]
First off, thank you for sharing this!

I'm a little confused, for some reason the output I'm getting looks exactly the same as before. It doesn't seem to be matching these new tings like dstip-private that you've added. 

here is my config 

#logstash.conf

input {
  udp {
    type => Sonicwall
	codec => plain {
       charset => "ISO-8859-1"
     }
    port => 514
  }
}

filter {
  if [type] == "sonicwall" {

    kv {
        exclude_keys => [ "<129>id", "<132>id", "<133>id","<134>id", "af_polid", "af_service", "app", "appid", "code", "fw", "m", "op", "sn" ]
    }
    
    date {
        match => [ "time", "yyyy-MM-dd HH:mm:ss z", "yyyy-MM-dd HH:mm:ss" ]
    }
    
    if [src] {
        grok {
            match => { 
                "src" => [
                    "%{IP:srcip}:%{INT:srcport}:%{DATA:srcint}:%{GREEDYDATA:srcname}",
                    "%{IP:srcip}:%{INT:srcport}:%{DATA:srcint}",
                    "%{IP:srcip}::%{DATA:srcint}",
                    ":%{INT:srcport}"
                ]  
            }
        }
    }
    
    if [dst] {
        grok {
            match => { 
                "dst" => [ 
                    "%{IP:dstip}:%{INT:dstport}:%{DATA:dstint}:%{GREEDYDATA:dstinfo}",
                    "%{IP:dstip}:%{INT:dstport}:%{DATA:dstint}",
                    "%{IP:dstip}::%{DATA:dstint}",
                    ":%{INT:dstport}"
                ]
            }
        }
    }
    
    # Sanitize fields with \r after recent firmware update
    mutate {
        gsub => [
            "sent", "\r", "",
            "rcvd", "\r", "",
            "cdur", "\r", "",
            "spkt", "\r", "",
            "rpkt", "\r", "",
            "proto", "\r", ""
        ]
    }
    
    # Assign network tags based on IP
    if [dstip] {
        cidr {
            add_tag => ["dstip-private"]
            address => ["%{dstip}"]
            network => ["10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16"]
        }                
        
        # Parse GeoIP info
        if "dstip-private" not in [tags] {
            geoip {
                source => "dstip"
                target => "dstip_geoip"
                fields => ["country_name", "real_region_name", "city_name", "location"]
            }
        }
    }
    
    # Replace srcname with srcip if srcname does not exist
    if ![srcname] and [srcip] {
        mutate {
            replace => { "srcname" => "%{srcip}" }
        }
    }	
    
    # Replace dstname with dstinfo or dstip if dstname does not exist
    if ![dstname] and [dstinfo] {
        mutate {
            replace => { "dstname" => "%{dstinfo}" }
        }
    } else if ![dstname] and [dstip] {
        mutate {
            replace => { "dstname" => "%{dstip}" }
        }
    }	
    
    mutate {
        lowercase => [ "msg", "appName", "sess", "fw_action", "srcint", "dstint", "Category"]
        remove_field => [ "src", "dst", "dstinfo", "message", "time" ]
        
    }	    
		
}
}

output {
        elasticsearch { hosts => ["localhost:9201"] }
	stdout { codec => rubydebug }
		}

and I get output that looks like this 

{
    "@timestamp" => 2016-11-11T15:54:22.742Z,
      "@version" => "1",
          "host" => "192.168.25.249",
       "message" => "<134>id=firewall sn=0017C5CA31XX time=\"2016-11-11 10:54:22\" fw=24.246.109.01 pri=6 c=1024 m=537 msg=\"Connection Closed\" app=11 sess=Auto n=137947643 usr=\"DOMAIN\\testuser\" src=192.168.26.56:56890:X0 dst=104.244.43.231:443:X1 proto=tcp/https sent=334 rcvd=191 spkt=6 rpkt=4 cdur=16",
          "type" => "Sonicwall"
}

**Comment by anhlqn** [comment]
The `"type => Sonicwall"` in your input block is different from `if [type] == "sonicwall"` in the filter block, so messages are not processed by the filter. I'd rather use lowercase for `type` across the config file.

**Comment by BustedSec** [comment]
ok, I changed it so it's lowercase but now it's throwing an error 

C:\LogStash\bin>logstash -f Logstash-SW.conf
Sending Logstash logs to C:/LogStash/logs which is now configured via log4j2.properties.
[2016-11-11T12:56:49,272][INFO ][logstash.inputs.udp      ] Starting UDP listener {:address=>"0.0.0.0:514"}
[2016-11-11T12:56:49,491][INFO ][logstash.outputs.elasticsearch] Elasticsearch pool URLs updated {:changes=>{:removed=>[], :added=>["http://localhost:9201"]}}
[2016-11-11T12:56:49,491][INFO ][logstash.outputs.elasticsearch] Using mapping template from {:path=>nil}
[2016-11-11T12:56:49,631][INFO ][logstash.outputs.elasticsearch] Attempting to install template {:manage_template=>{"template"=>"logstash-*", "version"=>50001, "settings"=>{"index.refresh_interval"=>"5s"}, "mappings"=>{"_default_"=>{"_all"=>{"enabled"=>true, "norms"=>false}, "dynamic_templates"=>[{"message_field"=>{"path_match"=>"message", "match_mapping_type"=>"string", "mapping"=>{"type"=>"text", "norms"=>false}}}, {"string_fields"=>{"match"=>"*", "match_mapping_type"=>"string", "mapping"=>{"type"=>"text", "norms"=>false, "fields"=>{"keyword"=>{"type"=>"keyword"}}}}}], "properties"=>{"@timestamp"=>{"type"=>"date", "include_in_all"=>false}, "@version"=>{"type"=>"keyword", "include_in_all"=>false}, "geoip"=>{"dynamic"=>true, "properties"=>{"ip"=>{"type"=>"ip"}, "location"=>{"type"=>"geo_point"}, "latitude"=>{"type"=>"half_float"}, "longitude"=>{"type"=>"half_float"}}}}}}}}
[2016-11-11T12:56:49,647][INFO ][logstash.outputs.elasticsearch] New Elasticsearch output {:class=>"LogStash::Outputs::ElasticSearch", :hosts=>["localhost:9201"]}
[2016-11-11T12:56:50,069][INFO ][logstash.filters.geoip   ] Using geoip database {:path=>"C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-filter-geoip-4.0.3-java/vendor/GeoLite2-City.mmdb"}
[2016-11-11T12:56:50,116][INFO ][logstash.pipeline        ] Starting pipeline {"id"=>"main", "pipeline.workers"=>8, "pipeline.batch.size"=>125, "pipeline.batch.delay"=>5, "pipeline.max_inflight"=>1000}
[2016-11-11T12:56:50,131][INFO ][logstash.pipeline        ] Pipeline main started
[2016-11-11T12:56:50,272][INFO ][logstash.agent           ] Successfully started Logstash API endpoint {:port=>9600}
[2016-11-11T12:56:55,381][ERROR][logstash.filters.geoip   ] Unknown error while looking up GeoIP data {:exception=>#<Exception: [real_region_name] is not a supported field option.>, :field=>"dstip", :event=>2016-11-11T17:56:55.000Z 192.168.25.249 <134>id=firewall sn=0017C5CA31D8 time="2016-11-11 12:56:55" fw=24.246.109.66 pri=6 c=1024 m=97 app=9 sess=Auto n=46826761 usr="DOMAIN\MMostransky" src=192.168.26.71:52690:X0 dst=54.236.3.173:80:X1 proto=tcp/http op=1 sent=462 rcvd=509 result=200 dstname=pubsub.pubnub.com arg=/v2/presence/sub_key/sub-c-9709e7f4-39f8-11e5-a5d7-02ee2ddab7fe/uuid/0141ff8e-1bb0-48c1-9bb6-e0a109ce2393?uuid code=64 Category="Not Rated"}
[2016-11-11T12:56:55,381][ERROR][logstash.pipeline        ] Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash. {"exception"=>#<Exception: [real_region_name] is not a supported field option.>, "backtrace"=>["C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-filter-geoip-4.0.3-java/lib/logstash/filters/geoip.rb:228:in `populate_geo_data'", "org/jruby/RubyArray.java:1613:in `each'", "C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-filter-geoip-4.0.3-java/lib/logstash/filters/geoip.rb:195:in `populate_geo_data'", "C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-filter-geoip-4.0.3-java/lib/logstash/filters/geoip.rb:161:in `filter'", "C:/LogStash/logstash-core/lib/logstash/filters/base.rb:156:in `multi_filter'", "org/jruby/RubyArray.java:1613:in `each'", "C:/LogStash/logstash-core/lib/logstash/filters/base.rb:153:in `multi_filter'", "C:/LogStash/logstash-core/lib/logstash/filter_delegator.rb:41:in `multi_filter'", "(eval):866:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):863:in `initialize'", "org/jruby/RubyProc.java:281:in `call'", "(eval):881:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):877:in `initialize'", "org/jruby/RubyProc.java:281:in `call'", "(eval):931:in `initialize'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):923:in `initialize'", "org/jruby/RubyProc.java:281:in `call'", "(eval):358:in `filter_func'", "C:/LogStash/logstash-core/lib/logstash/pipeline.rb:260:in `filter_batch'", "org/jruby/RubyProc.java:281:in `call'", "C:/LogStash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:186:in `each'", "org/jruby/RubyHash.java:1342:in `each'", "C:/LogStash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:185:in `each'", "C:/LogStash/logstash-core/lib/logstash/pipeline.rb:258:in `filter_batch'", "C:/LogStash/logstash-core/lib/logstash/pipeline.rb:246:in `worker_loop'", "C:/LogStash/logstash-core/lib/logstash/pipeline.rb:225:in `start_workers'"]}
Exception: [real_region_name] is not a supported field option.
  populate_geo_data at C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-filter-geoip-4.0.3-java/lib/logstash/filters/geoip.rb:228
               each at org/jruby/RubyArray.java:1613
  populate_geo_data at C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-filter-geoip-4.0.3-java/lib/logstash/filters/geoip.rb:195
             filter at C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-filter-geoip-4.0.3-java/lib/logstash/filters/geoip.rb:161
       multi_filter at C:/LogStash/logstash-core/lib/logstash/filters/base.rb:156
               each at org/jruby/RubyArray.java:1613
       multi_filter at C:/LogStash/logstash-core/lib/logstash/filters/base.rb:153
       multi_filter at C:/LogStash/logstash-core/lib/logstash/filter_delegator.rb:41
         initialize at (eval):866
               each at org/jruby/RubyArray.java:1613
         initialize at (eval):863
               call at org/jruby/RubyProc.java:281
         initialize at (eval):881
               each at org/jruby/RubyArray.java:1613
         initialize at (eval):877
               call at org/jruby/RubyProc.java:281
         initialize at (eval):931
               each at org/jruby/RubyArray.java:1613
         initialize at (eval):923
               call at org/jruby/RubyProc.java:281
        filter_func at (eval):358
       filter_batch at C:/LogStash/logstash-core/lib/logstash/pipeline.rb:260
               call at org/jruby/RubyProc.java:281
               each at C:/LogStash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:186
               each at org/jruby/RubyHash.java:1342
               each at C:/LogStash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:185
       filter_batch at C:/LogStash/logstash-core/lib/logstash/pipeline.rb:258
        worker_loop at C:/LogStash/logstash-core/lib/logstash/pipeline.rb:246
      start_workers at C:/LogStash/logstash-core/lib/logstash/pipeline.rb:225
[2016-11-11T12:56:55,553][WARN ][logstash.inputs.udp      ] UDP listener died {:exception=>#<Errno::EBADF: Bad file descriptor - Bad file descriptor>, :backtrace=>["org/jruby/RubyIO.java:3705:in `select'", "C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-input-udp-3.0.3/lib/logstash/inputs/udp.rb:77:in `udp_listener'", "C:/LogStash/vendor/bundle/jruby/1.9/gems/logstash-input-udp-3.0.3/lib/logstash/inputs/udp.rb:50:in `run'", "C:/LogStash/logstash-core/lib/logstash/pipeline.rb:331:in `inputworker'", "C:/LogStash/logstash-core/lib/logstash/pipeline.rb:325:in `start_input'"]}

I assume I must be missing a plugin

**Comment by anhlqn** [comment]
Remove this block to see if everything is parsed correctly
```
    # Parse GeoIP info
    if "dstip-private" not in [tags] {
        geoip {
            source => "dstip"
            target => "dstip_geoip"
            fields => ["country_name", "real_region_name", "city_name", "location"]
        }
    }
```
GeoIP is a standard filter plugin. What version of Logstash are you using?

**Comment by BustedSec** [comment]
I'm using version 5 - I installed it on a windows box, trying that now

**Comment by BustedSec** [comment]
also, thank you for your help

**Comment by anhlqn** [comment]
LS 5.0 may have breaking changes that I'm not aware of. My config is for LS 2.3.x and 2.4.x

**Comment by BustedSec** [comment]
ok, it's parsing with that section commented out. Something must be off with my geoip plugin.

**Comment by BustedSec** [comment]
do you have any kibana visualizations or dashboards you could share based off this? I'm new to the elastic stack and still learning wtf I'm doing.

**Comment by anhlqn** [comment]
Field names for geoip were changed. Try this
```
# Parse GeoIP info
    if "dstip-private" not in [tags] {
        geoip {
            source => "dstip"
            target => "dstip_geoip"
            fields => ["country_name", "region_name", "city_name", "location"]
        }
    }
```
`real_region_name` is now `region_name`
Full list of the geoip fields that you can export https://www.elastic.co/guide/en/logstash/5.0/plugins-filters-geoip.html#plugins-filters-geoip-fields

**Comment by BustedSec** [comment]
yup, that was it. 

works with 

      # Parse GeoIP info
        if "dstip-private" not in [tags] {
            geoip {
                source => "dstip"
                target => "dstip_geoip"
                fields => ["country_name", "region_name", "city_name", "location"]
            }
        }
    }

**Comment by anhlqn** [comment]
A few visualizations you can create:

- Throughput over Time based on rcvd and sent field (mapped as long type)
- Events over Time
- Top Visited Sites (dstname field)
- Top Users (usr field)
- Top Services/Protocols (http, https, dns, etc.)
- Web Traffic by Category (Category field)
- Web Traffic on the map (with dstip_geoip.location field)

**Comment by BustedSec** [comment]
very cool! 

I'm trying to do these now. Decided to start with the Web Traffic on the map but it's only letting me pick geoip.location from the dropdown and not showing me dstip_geo.location - I'll figure this out, this is so cool! thanks so much for your help!

**Comment by BustedSec** [comment]
can you give me an example of how you crated the web traffic on map? I'm having more trouble with it then I thought.

**Comment by anhlqn** [comment]
![sonicwall-traffic-on-map](https://cloud.githubusercontent.com/assets/3209539/20232384/daf2cd62-a81c-11e6-90b9-017583d77986.jpg)

The geoip_location should be mapped correctly in Elasticsearch. Things change a lot in Elastic stack 5.0

**Comment by BustedSec** [comment]
How do you map geoip_location in ElasticSearch? which file do you edit?

**Comment by anhlqn** [comment]
@BustedSec https://www.elastic.co/guide/en/elasticsearch/reference/2.4/mapping.html

**Comment by BustedSec** [comment]
Can you dumb it down for me, I undertstand the syntax there but have no idea how to do this 
"PUT" command - I installed curl on windows but am still lost.

**Comment by anhlqn** [comment]
On Kibana 5, you can open the Dev/Console app, and run `GET _template` to see existing templates in Elasticsearch. Then you can look at the Logstash template to see how fields are mapped.

`PUT template/sonicwall {}` with some mapping is used to add a new mapping named `sonicwall`

**Comment by BustedSec** [comment]
so would something like this work?

PUT template/sonicwall 
	{
          "dstip_geoip.location": {
            "dynamic": true,
            "properties": {
              "ip": {
                "type": "ip"
              },
              "latitude": {
                "type": "half_float"
              },
              "location": {
                "type": "geo_point"
              },
              "longitude": {
                "type": "half_float"
              }
            }

**Comment by BustedSec** [comment]
because when I try that I get No handler found for uri [/template/sonicwall] and method [PUT]

**Comment by anhlqn** [comment]
Below is my current template that goes with the Logstash filter
```
PUT _template/sonicwall
{
  "order": 0,
  "template": "sonicwall*",
  "settings": {
    "index": {
      "refresh_interval": "15s",
      "number_of_shards": "1",
      "number_of_replicas": "1"
    }
  },
  "mappings": {
    "_default_": {
      "include_in_all": false,
      "dynamic_templates": [
        {
          "string_fields": {
            "mapping": {
              "index": "not_analyzed",
              "omit_norms": true,
              "type": "string"
            },
            "match_mapping_type": "string",
            "match": "*"
          }
        }
      ],
      "properties": {
        "cdur": {
          "type": "long"
        },
        "dstname": {
          "include_in_all": true,
          "index": "not_analyzed",
          "type": "string"
        },
        "srcip": {
          "include_in_all": true,
          "index": "not_analyzed",
          "type": "string"
        },
        "rpkt": {
          "type": "long"
        },
        "spkt": {
          "type": "long"
        },
        "srcname": {
          "include_in_all": true,
          "index": "not_analyzed",
          "type": "string"
        },
        "usr": {
          "include_in_all": true,
          "index": "not_analyzed",
          "type": "string"
        },
        "arg": {
          "index": "analyzed",
          "type": "string"
        },
        "dstip_geoip": {
          "dynamic": "true",
          "properties": {
            "location": {
              "type": "geo_point"
            }
          }
        },
        "rcvd": {
          "type": "long"
        },
        "dstip": {
          "include_in_all": true,
          "index": "not_analyzed",
          "type": "string"
        },
        "sent": {
          "type": "long"
        }
      }
    }
  },
  "aliases": {

  }
}
```
Change `"template": "sonicwall*",` to match your index naming pattern. This `"template": "sonicwall*",` is for indexes like sonicwall-2016.44 (weekly) or sonicwall-2016.11.18 (daily)

You may also need to delete old sonicwall indexes so that there won't be field mapping conflicts.

**Comment by BustedSec** [comment]
Thank you for that, I screwed stuff up by adding the x-tools plugin so I could get that console, and now I can't authenticate to it using my logstash.conf - it can't parse that file now that I've followed the guide here https://www.elastic.co/guide/en/x-pack/current/logstash.html so I'm working on fixing the authentication problem. I wiped my old index and the indecies that were stored. I'm going to keep working on this - thank you so much for sharing those configs - they've made things a lot easier.

**Comment by BustedSec** [comment]
ok, one issue down - fixed that issue by disabling authentication in the kibana.yml and elasticsearch.yml file with xpack.security.enabled: false

now I am seeing dstip_geoip.location in the index but it's set as a number instead of a geopoint - trying to figure out how to change that since I can't edit it there

**Comment by anhlqn** [comment]
Could you run these two queries and paste the output here
```
GET <your_index>/_search
GET <your_index>/_mapping
```
If you see something like `"location": [ -71.34, 41.12 ]`, then it's normal.
More info https://www.elastic.co/guide/en/elasticsearch/reference/current/geo-point.html

**Comment by anhlqn** [comment]
I wrote a tutorial about ELK stack a while ago. It's about processing Netflow, but the process should be similar to other kinds of log http://www.secureict.info/2015/11/process-netflow-with-nprobe-and.html
